x/staking/keeper/validator.go

```diff
@@ -247,68 +240,114 @@ func (k Keeper) GetValidators(ctx sdk.Context, maxRetrieve uint16) (validators [
+	return
+}
+
+// Set the last validator power.
+func (k Keeper) SetLastValidatorPower(ctx sdk.Context, operator sdk.ValAddress, power int64) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(power)
+	store.Set(GetLastValidatorPowerKey(operator), bz)
+}
+
+// Delete the last validator power.
+func (k Keeper) DeleteLastValidatorPower(ctx sdk.Context, operator sdk.ValAddress) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(GetLastValidatorPowerKey(operator))
+}
+
 // returns an iterator for the consensus validators in the last block
 func (k Keeper) LastValidatorsIterator(ctx sdk.Context) (iterator sdk.Iterator) {
 	store := ctx.KVStore(k.storeKey)
 	iterator = sdk.KVStorePrefixIterator(store, LastValidatorPowerKey)
 	return iterator
 }
 
-// get the current group of bonded validators sorted by power-rank
-func (k Keeper) GetBondedValidatorsByPower(ctx sdk.Context) []types.Validator {
+// Iterate over last validator powers.
+func (k Keeper) IterateLastValidatorPowers(ctx sdk.Context, handler func(operator sdk.ValAddress, power int64) (stop bool)) {
+	store := ctx.KVStore(k.storeKey)
+	iter := sdk.KVStorePrefixIterator(store, LastValidatorPowerKey)
+	defer iter.Close()
+	for ; iter.Valid(); iter.Next() {
+		addr := sdk.ValAddress(iter.Key()[len(LastValidatorPowerKey):])
+		var power int64
+		k.cdc.MustUnmarshalBinaryLengthPrefixed(iter.Value(), &power)
+		if handler(addr, power) {
+			break
+		}
+	}
+}
+
+// get the group of the bonded validators
+func (k Keeper) GetLastValidators(ctx sdk.Context) (validators []types.Validator) {
 	store := ctx.KVStore(k.storeKey)
+
+	// add the actual validator power sorted store
 	maxValidators := k.MaxValidators(ctx)
-	validators := make([]types.Validator, maxValidators)
+	validators = make([]types.Validator, maxValidators)
 
-	iterator := sdk.KVStoreReversePrefixIterator(store, ValidatorsByPowerIndexKey)
+	iterator := sdk.KVStorePrefixIterator(store, LastValidatorPowerKey)
 	defer iterator.Close()
 
 	i := 0
-	for ; iterator.Valid() && i < int(maxValidators); iterator.Next() {
-		address := iterator.Value()
-		validator := k.mustGetValidator(ctx, address)
+	for ; iterator.Valid(); iterator.Next() {
 
-		if validator.Status == sdk.Bonded {
-			validators[i] = validator
-			i++
+		// sanity check
+		if i >= int(maxValidators) {
+			panic("more validators than maxValidators found")
 		}
+		address := AddressFromLastValidatorPowerKey(iterator.Key())
+		validator := k.mustGetValidator(ctx, address)
+
+		validators[i] = validator
+		i++
 	}
 	return validators[:i] // trim
 }
 
-// returns an iterator for the current validator power store
-func (k Keeper) ValidatorsPowerStoreIterator(ctx sdk.Context) (iterator sdk.Iterator) {
-	store := ctx.KVStore(k.storeKey)
-	iterator = sdk.KVStoreReversePrefixIterator(store, ValidatorsByPowerIndexKey)
-	return iterator
-}
+//_______________________________________________________________________
```

Comment:
delete this line?

Reply:
I like it as a visual queue, the next three functions are related (a group)... a smaller division rather then adding a whole new file

Reply:
ok nw